man/fr/armadito-scan.1.txt
ARMADITO-SCAN(1)
================
:doctype: manpage

NOM
---
armadito-scan - outil de scan en ligne de commande


SYNOPSIS
--------
*armadito-scan* ['OPTIONS'] 'FILE|DIR'


DESCRIPTION
-----------
La commande *armadito-scan*(1) lance une recherche de malwares dans le fichier ou répertoire passé en argument et affiche les résultats sur sa sortie standard.

Cette commande utilise la librairie de scan *libarmadito*(3).


OPTIONS
-------
*-r, --recursive*::
    Récursif.
    Analyse récursive des répertoires. Lance la recherche dans le répertoire et dans tous ses sous-répertoires.

*-t, --threaded*::
    Threadé.
    Utilise un ensemble de threads pour accélérer l'analyse.

*-n, --no-summary*::
    Pas de résumé.
    N'affiche pas le résumé de l'analyse à la fin de l'exécution.

*-h, --help*::
    Aide
    Affiche l'aide et termine l'exécution.


EXIT STATUS
-----------
*0*::
    Succès

*1*::
    Erreur


AUTHOR
------
armadito-scan a été écrit par François Déchelle.


RESOURCES
---------
Site web: <http://armadito.org/>


COPYING
-------
Copyright \(C) 2014 qui?. Free use of this software is
granted under the terms of the GNU General Public License (GPL).


gui/ChangeLog
2015-09-23  François  <francois@dechelle.net>

	* fixed crash if 'cancel' pressed when selecting directory to scan

2015-09-14  François  <francois@dechelle.net>

	* trying to fix analysis file selection box; cannot select file and directories apparently

	* fixed compilation; works; commented out 'Recent analysis' for now

2015-07-31  François  <francois@dechelle.net>

	* fixed i18n initialization problem

2015-05-29  François  <francois@dechelle.net>

	* implementation of recent scans in progress; almost works

2015-05-28  François  <francois@dechelle.net>

	* adding recent scan menu

2015-05-27  François  <francois@dechelle.net>

	* internationalization completed

2015-05-12  François  <francois@dechelle.net>

	* dbus debug in progress

2015-04-02  François  <francois@dechelle.net>

	* added scan duration in scan dialog

	* finished i18n; replaced systray menu creation with manual menu, no more XML or glade

2015-03-27  François  <francois@dechelle.net>

	* added i18n using intltool; only language added yet is fr

2015-03-16  François  <francois@dechelle.net>

	* added systray implementation with old gtk_status_icon()

	* added non-scalable PNG icons

2015-03-12  François  <francois@dechelle.net>

	* fixed icon path

2015-03-11  François  <francois@dechelle.net>

	* moved interface to GTK; Qt is phased out

2015-03-10  François  <francois@dechelle.net>

	* added about dialog

2015-03-06  François  <francois@dechelle.net>

	* added systray icon and menu; works almost like Qt version (maybe better...)

	* adding file selection dialog

	* fixed state, action & path display

	* fixed some sorting and resizing problems

2015-03-05  François  <francois@dechelle.net>

	* fixed close button

	* added unhandled file count

	* added current scanning path displayed in progress bar

	* fixed progress bar and to scan file count

	* changed implementation of GTK model to use polled I/O instead of threads; works!

2015-02-04  François  <francois@dechelle.net>

	* added implementation of model counters using glib/gobject

2015-02-03  François  <francois@dechelle.net>

	* starting implementation of GTK gui

2015-01-29  François  <francois@dechelle.net>

	* moving Qt gui to its own directory; preparing move to gtk ;-)

2015-01-13  François  <francois@dechelle.net>

	* trying to fix report table column resizing; added new counters

2015-01-08  François  <francois@dechelle.net>

	* fixed configure.ac and systray.cc w.r.t. CentOS packaging

2014-12-26  François  <francois@dechelle.net>

	* removed -rpath option in src/Makefile.am, which made lintian fail

2014-12-10  François  <francois@dechelle.net>

	* fixed make dist

	* added a 'About' menu entry, with associated dialog

2014-12-05  François  <francois@dechelle.net>

	* added a 'Recent scans' menu entry, and fixed initialization of ScanWidget when displaying a passed scan

2014-12-04  François  <francois@dechelle.net>

	* changed scan window size and position

2014-12-03  François  <francois@dechelle.net>

	* added a close button to the scan window, enabled when scan is completed

 	* tested umwsu.cpp w.r.t. load thread

	* changing umwsu.cpp in order to call umwsu_open at startup in a separate thread, in order not to block interface startup

2014-12-01  François  <francois@dechelle.net>

	* added .desktop file; added make install rules for .desktop file and icons

2014-11-25  François  <francois@dechelle.net>

	* improved model to display analysis file names and select them (copy/paste works)

	* temporarily giving up with icon animation; will see later

2014-11-24  François  <francois@dechelle.net>

	* changed scan model to call umwsu_*() inside a QThread; now tables updates continuously

	* implemented correctly the model derived from QAbstractTableModel; now the table updates ok

2014-11-19  François  <francois@dechelle.net>

	* adding QAbstractTableModel derived model for scan report

2014-11-18  François  <francois@dechelle.net>

	* switched to uhuru-linux-gui git repo

	* fixed some compilation and installation problems with automake/autoconf; now runs ok with icon

2014-11-17  François  <francois@dechelle.net>

	* compile and run with automake/autoconf; will make packaging much easier

	* added automake/autoconf stuff

2014-11-05  François  <francois@dechelle.net>

	* changed scan widget components, changed constructors signatures

	* first working version with systray; scans ok

	* added file selection dialog to systray action

	* added systray to IHM

	* reorganized sources; moved origin UI to legacy/

2014-09-24  François Déchelle  <francois@dechelle.net>

	* ported Uhuru/ to QT4: works ok, except strange compile error in scanondemand.cpp




tools/armadito-scan.c
/***

Copyright (C) 2015, 2016 Teclib'

This file is part of Armadito core.

Armadito core is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Armadito core is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with Armadito core.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "armadito-config.h"

#include <net/netdefaults.h>
#include <net/unixsockclient.h>
#include <rpc/io.h>
#include <rpc/rpctypes.h>
#include <libjrpc/jrpc.h>

#include <assert.h>
#include <getopt.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define PROGRAM_NAME "armadito-scan"
#define PROGRAM_VERSION PACKAGE_VERSION

struct scan_options {
	int recursive;
	int threaded;
	int no_summary;
	int print_clean;
	const char *path_to_scan;
	int verbose;
	const char *unix_socket_path;
};

static struct option scan_option_defs[] = {
	{"help",         no_argument,        0, 'h'},
	{"version",      no_argument,        0, 'V'},
	{"verbose",      no_argument,        0, 'v'},
	{"recursive",    no_argument,        0, 'r'},
	{"threaded",     no_argument,        0, 't'},
	{"no-summary",   no_argument,        0, 'n'},
	{"socket-path",  required_argument,  0, 'a'},
#if O
	{"print-clean",  no_argument,        0, 'c'},
#endif
	{0, 0, 0, 0}
};

static void version(void)
{
	printf("%s %s\n", PROGRAM_NAME, PROGRAM_VERSION);
	exit(EXIT_SUCCESS);
}

static void usage(void)
{
	fprintf(stderr, "usage: armadito-scan [options] FILE|DIR\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Armadito antivirus scanner\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Options:\n");
	fprintf(stderr, "  --help  -h                    print help and quit\n");
	fprintf(stderr, "  --version -V                  print program version\n");
	fprintf(stderr, "  --verbose -v                  print HTTP trafic\n");
	fprintf(stderr, "  --recursive  -r               scan directories recursively\n");
	fprintf(stderr, "  --threaded -t                 scan using multiple threads\n");
	fprintf(stderr, "  --no-summary -n               disable summary at end of scanning\n");
	fprintf(stderr, "  --socket-path=PATH | -a PATH  unix socket path (default is " DEFAULT_SOCKET_PATH ")\n");
#if O
	/* yet not available with rpc api */
	fprintf(stderr, "  --print-clean -c              print also clean files as they are scanned\n");
#endif
	fprintf(stderr, "\n");

	exit(-1);
}

static void parse_options(int argc, char **argv, struct scan_options *opts)
{
	opts->recursive = 0;
	opts->threaded = 0;
	opts->no_summary = 0;
	opts->print_clean = 0;
	opts->path_to_scan = NULL;
	opts->unix_socket_path = DEFAULT_SOCKET_PATH;
	opts->verbose = 0;

	while (1) {
		int c, option_index = 0;

		c = getopt_long(argc, argv, "hVvrtna:", scan_option_defs, &option_index);

		if (c == -1)
			break;

		switch (c) {
		case 'h': /* help */
			usage();
			break;
		case 'V': /* version */
			version();
			break;
		case 'v': /* verbose */
			opts->verbose = 1;
			break;
		case 'r': /* recursive */
			opts->recursive = 1;
			break;
		case 't': /* threaded */
			opts->threaded = 1;
			break;
		case 'n': /* no-summary */
			opts->no_summary = 1;
			break;
		case 'a': /* path */
			opts->unix_socket_path = strdup(optarg);
			break;
		case '?':
			/* getopt_long already printed an error message. */
			break;
#if 0
		case 'c': /* print-clean */
			opts->print_clean = 1;
			break;
#endif
		default:
			abort ();
		}
	}

	if (optind != argc - 1)
		usage();

	opts->path_to_scan = strdup(argv[optind]);
}

static void scan_cb(json_t *result, void *user_data)
{
	int ret;

	*(int *)user_data = 1;
}

static int do_scan(struct scan_options *opts)
{
	struct jrpc_connection *conn;
	int client_sock;
	int *p_client_sock;
	int ret;
	struct a6o_rpc_scan_param param;
	json_t *j_param;
	static int done = 0;

	client_sock = unix_client_connect(opts->unix_socket_path, 10);

	if (client_sock < 0) {
		perror("cannot connect");
		exit(EXIT_FAILURE);
	}

	conn = jrpc_connection_new(NULL, NULL);

	p_client_sock = malloc(sizeof(int));
	*p_client_sock = client_sock;

	jrpc_connection_set_read_cb(conn, unix_fd_read_cb, p_client_sock);
	jrpc_connection_set_write_cb(conn, unix_fd_write_cb, p_client_sock);

	/* jrpc_connection_set_error_handler(conn, client_error_handler); */

	param.root_path = opts->path_to_scan;
	if ((ret = JRPC_STRUCT2JSON(a6o_rpc_scan_param, &param, &j_param)))
		return ret;

	if ((ret = jrpc_call(conn, "scan", j_param, scan_cb, &done)))
		return ret;

	while((ret = jrpc_process(conn)) != JRPC_EOF && !done)
		;

	if (close(client_sock) < 0)
		perror("closing connection");

	return 0;
}

int main(int argc, char **argv)
{
	struct scan_options *opts = (struct scan_options *)malloc(sizeof(struct scan_options));

	parse_options(argc, argv, opts);

	do_scan(opts);

	return 0;
}
